src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -23,10 +23,18 @@ namespace crimson::osd {
 void ClientRequest::Orderer::requeue(Ref<PG> pg)
 {
   LOG_PREFIX(ClientRequest::Orderer::requeue);
-  for (auto &req: list) {
-    DEBUGDPP("requeueing {}", *pg, req);
-    req.reset_instance_handle();
-    std::ignore = req.with_pg_process(pg);
+  std::list<ClientRequest*> to_requeue;
+  for (auto &req : list) {
+    to_requeue.emplace_back(&req);
+  }
+  // Client requests might be destroyed in the following
+  // iteration leading to short lived dangling pointers
+  // to those requests, but this doesn't hurt as we won't
+  // dereference those dangling pointers.
+  for (auto req: to_requeue) {
+    DEBUGDPP("requeueing {}", *pg, *req);
```

Comment:
In continuation to the previous comment, will checking `if (req)` before actually re-queuing help us to avoid holding the additional reference (`to_requeue`)?

Reply:
As discussed in https://github.com/ceph/ceph/pull/58708#discussion_r1694599996, this may not help. Because after the removal, that `req` would have become a dangling pointer which is not NULL, so the check `if (req)` would succeed.

Reply:
`with_pg_process` shouldn't actually be executed until `when_all_succeed` is called, right?
We just pack the future in a vector without executing it until the loop iteration is finished.

Reply:
> `with_pg_process` shouldn't actually be executed until `when_all_succeed` is called, right? We just pack the future in a vector without executing it until the loop iteration is finished.

Um, I don't think so. I think `with_pg_process` is executed the moment it's called just like conventional method invocation, it's just that it may not give the caller what it wants when `with_pg_process` returns. However, if the pg is active and the request should be discarded, there wouldn't be any blocking conditions that prevent `with_pg_process` from getting the result, so the process of the request would finish when it returns.

Reply:
> Um, I don't think so. I think `with_pg_process` is executed the moment it's called just like conventional method invocations, it's just that it may not give the caller what it wants when `with_pg_process` returns. However, if the pg is active and the request should be discarded, there wouldn't be any blocking conditions that prevent `with_pg_process` from getting the result, so the process of the request would finish when it returns.

Ack, you are right.